### ISSUE TITLE
Imperative API: Improve ability to read/write local store data

### ISSUE THREAD

**Issue by serle** [apollographql/apollo-client#1876]
Instead of using `client.readFragment()` and `client.writeFragment()` with all their GraphQL like complexity, one could operate against the cache in a more array-like way i.e. using concepts such as filters and maps.  This API would be limited to normalized cache data i.e. it operates under the assuming that Apollo has already normalized the cache by `__typename`. Thus this API would not be usable for GraphQL types that don't meet this assumption i.e. don't include an `id` and are therefore only materialized as embedded objects. 

The API could have signatures such as: 

`record_arr = client.readData(typename, filter_fn)` which returns an array of plain objects that can be modified in a normal way. 

Local mutations could take the form of:

`client.updateData(typename, record_arr)`, which joins by id and overwrites existing cache data
`client.createData(typename, record_arr)`, which creates new records in the `__typename` cache 
`client.deleteData(typename, record_arr)`. which marks existing cache records for delete

An API for the specification of the synchronization of these updates between the cache and the GraphQL server will be discussed in another issue, but the idea would be to setup these mutations  per `__typename` more centrally.

**Comment by jbaxleyiii** [comment]
@serle thank you for the feature request! We are working on a new store design and I think these ideas would be great to include in some manner.

Do you have a couple use cases to show how these could be used in practice? Thanks!

@shadaj notes for the new store implementation

**Comment by serle** [comment]
I have a scheduling system that queries scheduling data along with multiple sets of reference data. e.g. resources. The reference data can either be embedded in the query or materialized on the client through separate queries. Either way I end up with a scheduling data cache and multiple reference data caches as Apollo normalizes by `__typename`. I then operate against the scheduling data using client side scheduling algorithms that may make multiple record updates in response to user actions. Here are a couple of use cases:

1) I need to build a tree control so I need to augment each object with `__typename === schedule_item` with an `is_expanded` and an `is_visible` flag. At the moment Apollo does not allow one to change the shape of an object in the cache. Work around augment server-side. Down side is unnecessary transport cost.

2) I need to create an Undo/RedoManager on the client side as the scheduling algorithm can make multiple updates as a result of relatively minor user changes. I only want to propagate these once the user is sure that they should be saved to the server. I also want to allow the user to undo these updates in bulk and be able to cancel all updates. Having access to an `is_dirty` and a `last_txn_id` flag as well as filter based cache queries could assist in implementing and Undo/RedoManager.

3) Once the user is ready to save, I would like to instruct Apollo to save all objects that are dirty (or meet certain criteria e.g. have a give last_txn_id) in its cache for a given `__typename`. 

Hope this helps

**Comment by Fxlr8** [comment]
#1947 
This problem can also be solved by the ability of reading/writing local store.
